lib/chef/provider/execute.rb

```diff
@@ -39,7 +42,11 @@ def whyrun_supported?
       end
 
       def define_resource_requirements
-         # @todo: this should change to raise in some appropriate major version bump.
+        # @todo: this should change to raise in some appropriate major version bump.
+        requirements.assert(:all_actions) do |a|
+          a.assertion { validate_identity(new_resource.user, new_resource.password, new_resource.domain) }
+        end
```

Comment:
yeah the shape of this looks pretty wrong.  you're not using either the why-run text or the exception to raise in not-why-run.  the conditional has the side effect of doing all the raises, and will run in both why-run and not-why-run the same, which means the whole define_resource_requirements API is not really being used and could be dropped.

since i don't think that validate_identity relies on doing user lookups and is static analysis of the parameters, though, that means it should really be moved to the resource (i don't think any of the syntax checks there could depend upon prior resources being run unless i missed something).  in a few cases the assertion is simple ("on unix, if password is specified you must raise") which are easy to do directly from the property declaration.  for the ones where you need collaboration ("if user is set on windows, password must also be set") then that needs to be done in the after_created block of the resource.

if there is anything which depends on the user actually being created then that would need to go into define_resource_requirements, but then you'd need to have the full define_resource_requirements API here with the exception to raise, and the text to produce in why-run mode.